### PR TITLE
Feature/ATG-33 Next figure

### PR DESCRIPTION
**Pull request by IraCherginyshka**
create board model
update gameService
logic nextFigure component

### CODE REVIEW

File: src/app/components/game-board/game-board.component.ts [hunk 1]
```diff
@@ -96,22 +86,24 @@ export class GameBoardComponent implements OnInit, OnDestroy {
   private setInitialBoardState(): void {
     this.lineWithFigure = 0;
     this.figurePosition = CENTRAL_ITEM;
-    this.lineWithFigure = 0;
-    this.currentFigure = FigureModel.getRandomFigure();
+    this.currentFigure = this.gameService.currentFigure;
```
Comment (oleksandrstarov) [inline src/app/components/game-board/game-board.component.ts]: it is better practice to use function to some property from service and make property private

File: src/app/components/game-next-figure/game-next-figure.component.ts [hunk 2]
```diff
@@ -1,8 +1,40 @@
...
+  @ViewChild('nextCanvas', { static: true }) private nextCanvas: ElementRef<HTMLCanvasElement>;
+  private ctx: CanvasRenderingContext2D;
+  private nextFigure: FiguresColors[][];
+  private subscriptionNext: Subscription;
+
+  constructor(private gameService: GameService) {}
+
+  ngOnInit(): void {
+    this.ctx = this.nextCanvas.nativeElement.getContext('2d');
+    this.subscriptionNext = this.gameService.onUpdateFigures().subscribe(() => {
```
Comment (oleksandrstarov) [inline src/app/components/game-next-figure/game-next-figure.component.ts]: .onUpdateFigures() is not the best naming >> onNewFigureCreated()
and it is better to pass figure matrix via .next() so we catch this in subscription

File: src/app/models/block.model.ts [hunk 3]
```diff
@@ -1,13 +1,20 @@
...
+import { FiguresColors } from '../enums/figures-colors.enum';
 
 export class BlockModel {
   constructor(private ctx: CanvasRenderingContext2D, color: string) {
     this.ctx.fillStyle = color;
     this.ctx.strokeStyle = BORDER_COLOR;
   }
 
-  fillBoardBlock(x: number, y: number): void {
+  fillBoardBlock(x: number, y: number, nextFigure: boolean): void {
```
Comment (oleksandrstarov) [inline src/app/models/block.model.ts]: nextFigure naming should be isNextFigure for boolean

File: src/app/models/block.model.ts [hunk 4]
```diff
@@ -1,13 +1,20 @@
...
     this.ctx.fillRect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE);
-    this.ctx.strokeRect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE);
+    if (nextFigure) {
+      if (this.ctx.fillStyle !== FiguresColors.DEFAULT) {
+        this.ctx.strokeRect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE);
+      }
+    } else {
+      this.ctx.strokeRect(x * BLOCK_SIZE, y * BLOCK_SIZE, BLOCK_SIZE, BLOCK_SIZE);
+    }
```
Comment (oleksandrstarov) [inline src/app/models/block.model.ts]: maybe 
 if (!nextFigure || (nextFigure && this.ctx.fillStyle !== FiguresColors.DEFAULT)) {
?

File: src/app/models/block.model.ts (hunk 3 above)
Comment (oleksandrstarov) [inline src/app/models/block.model.ts]: Should be public?

File: src/app/models/block.model.ts [hunk 5]
```diff
@@ -1,13 +1,20 @@
 import { BLOCK_SIZE, BORDER_COLOR } from '../constants/board-component.const';
+import { FiguresColors } from '../enums/figures-colors.enum';
 
 export class BlockModel {
   constructor(private ctx: CanvasRenderingContext2D, color: string) {
```
Comment (oleksandrstarov) [inline src/app/models/block.model.ts]: Well, lets try to use Abstract class with default fillBoardBlock function and separate fill stroke one for GameBlockModel and other for NextBlockModel, what do you think, Iryna?

File: src/app/models/board.model.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  constructor(private ctx: CanvasRenderingContext2D) {}
+
+  static makeBoardEmptyMatrix(width: number, height: number): FiguresColors[][] {
+    return new Array(height).fill(new Array(width).fill(FiguresColors.DEFAULT));
+  }
+
+  public drawBoard(matrix: FiguresColors[][], nextFigure: boolean): void {
+    matrix.forEach((line, indexY) => {
+      line.forEach((item, indexX) =>
+        new BlockModel(this.ctx, item).fillBoardBlock(indexX, indexY, nextFigure),
```
Comment (oleksandrstarov) [inline src/app/models/board.model.ts]: With classes for Blocks we can only pass isNextFigure to this function and create appropriate model in loop

File: src/app/models/board.model.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import { BlockModel } from './block.model';
+
+export class BoardModel {
+  constructor(private ctx: CanvasRenderingContext2D) {}
+
+  static makeBoardEmptyMatrix(width: number, height: number): FiguresColors[][] {
+    return new Array(height).fill(new Array(width).fill(FiguresColors.DEFAULT));
+  }
+
+  public drawBoard(matrix: FiguresColors[][], nextFigure: boolean): void {
```
Comment (oleksandrstarov) [inline src/app/models/board.model.ts]: nextFigure >> isNextFigure

File: src/app/services/game.service.ts [hunk 8]
```diff
@@ -2,12 +2,17 @@ import { Injectable } from '@angular/core';
...
 
 @Injectable({ providedIn: 'root' })
 export class GameService {
   public isPlaying = false;
+  public currentFigure = FigureModel.getRandomFigure();
+  public nextFigure = FigureModel.getRandomFigure();
+
   private movementSubject = new Subject<FiguresMovement>();
   private gameStateSubject = new Subject<GameState>();
+  private nextFigureSubject = new Subject<void>();
```
Comment (oleksandrstarov) [inline src/app/services/game.service.ts]: Lets use FigureModel instead of void

**Review by oleksandrstarov — CHANGES_REQUESTED** [review]

File: src/app/services/game.service.ts (hunk 8 above)
Comment (IraCherginyshka) [reply src/app/services/game.service.ts]: Done

File: src/app/models/board.model.ts (hunk 7 above)
Comment (IraCherginyshka) [reply src/app/models/board.model.ts]: Done

File: src/app/components/game-board/game-board.component.ts (hunk 1 above)
Comment (IraCherginyshka) [reply src/app/components/game-board/game-board.component.ts]: Done

File: src/app/components/game-next-figure/game-next-figure.component.ts (hunk 2 above)
Comment (IraCherginyshka) [reply src/app/components/game-next-figure/game-next-figure.component.ts]: Done

File: src/app/models/block.model.ts (hunk 3 above)
Comment (IraCherginyshka) [reply src/app/models/block.model.ts]: Changed

File: src/app/models/block.model.ts (hunk 4 above)
Comment (IraCherginyshka) [reply src/app/models/block.model.ts]: Yes. Changed

File: src/app/models/block.model.ts (hunk 3 above)
Comment (IraCherginyshka) [reply src/app/models/block.model.ts]: No
Changed

File: src/app/models/board.model.ts (hunk 6 above)
Comment (IraCherginyshka) [reply src/app/models/board.model.ts]: Fixed

File: src/app/models/block.model.ts (hunk 5 above)
Comment (IraCherginyshka) [reply src/app/models/block.model.ts]: Changed
I hope

File: src/app/models/block.model.ts [hunk 9]
```diff
@@ -1,13 +1,12 @@
...
 
-export class BlockModel {
-  constructor(private ctx: CanvasRenderingContext2D, color: string) {
+export abstract class BlockModel {
+  constructor(protected ctx: CanvasRenderingContext2D, private color: string) {
     this.ctx.fillStyle = color;
     this.ctx.strokeStyle = BORDER_COLOR;
   }
 
   fillBoardBlock(x: number, y: number): void {
```
Comment (oleksandrstarov) [inline src/app/models/block.model.ts:9]: We can keep implementation in abstract class and create additional abstract fillStroke function which will be implemented in child classes

**Review by oleksandrstarov — CHANGES_REQUESTED** [review]

**Review by oleksandrstarov — APPROVED** [review]

**Pull request merged**